tests/macros/utils.js

```diff
@@ -154,11 +158,44 @@ function afterEachMacro(teardown) {
     });
 }
 
+/**
+ * This function validates its input as HTML. By default, it assumes the input
+ * is an HTML fragment, wrapping it to make a complete HTML document, but the
+ * second argument can be set to false to avoid wrapping. It returns null on
+ * success, and, on failure, a string detailing all of the errors.
+ *
+ * @param {string} html
+ * @param {boolean} fragment
+ */
+function lintHTML(html, fragment=true) {
+    if (fragment) {
+        html = `<!DOCTYPE html>
+                <html>
+                <head><title>test</title></head>
+                <body>${html}</body>
+                </html>`;
+    }
+    try {
+        execSync(`java -jar ${vnu} --errors-only --format text -`, {
```

Comment:
How long does it take to fire up the java process? If you start writing a lot of tests that use it, is it going to make the test suite slow? 

Is there a way to set this up to create the process on the first call and then keep it open and reuse it for subsequent calls? (I'm guessing that would work even though Jest runs tests in parallel... presumably there would be one Java process for each of Jest's child execution contexts, though I'm not really sure how that works.)

Not something to block on, but just wondering about the future of this.

I think you made the right decision keeping this synchronous... If it was async, you might end up with a bunch of tests all trying to start up java processes at the same time.

Reply:
Thanks for recording your thoughts on this. I had thought of these issues as well, and come to the same conclusion. I suspect `lintHTML()` will be used heavily, so it's something we may have to consider soon. I've used [Python's built-in process pool](https://docs.python.org/2/library/multiprocessing.html#using-a-pool-of-workers) for things like this, and wondered if Node.js had something similar. I saw [this](https://www.npmjs.com/package/workerpool) npm package that looked promising.

Reply:
> How long does it take to fire up the java process?

When vnu.jar is invoked from the command line, it takes about 2 seconds in total to start up the java process and then load the HTML5 schema and other init steps.

> Is there a way to set this up to create the process on the first call and then keep it open and reuse it for subsequent calls?

There’s no way to do that when invoking vnu.jar for command-line checking. There is a way to do it by instead starting the validator as a persistent web server, in which case there’s an HTTP/REST API that can be used to send documents to it one-by-one for checking and get the results back.

That avoids paying the java/init startup cost for each document request but has other tradeoffs — including that the exit code doesn’t automatically get set the way it does when invoking vnu.jar from the command-line (that is, if the validator finds any errors in a document when you check it from the command-line, when the process exits, it exits with a non-zero exit status — whereas if the document has no errors, it exits zero).

However, since, when invoking vnu.jar for command-line-checking, you can send documents to it in batches, you can also avoid paying the java/init startup cost by queuing/buffering up the documents and then feeding them to vnu.jar just once as a set.

But I guess that alternative may not be a good fit for the test harness here.